Kp 7481 sync records

kaisakui opened the pull request:
This PR enables deleting records from Metax based on if they are "extra" compared to those in Metashare.
Changes include:
- Refactoring Metax to a class
- GET request that fetches all records in a catalog and returns their PIDs in a list
- Comparison functionality that spits out those PIDs not existing in Metashare but only in Metax (anymore, since we are assuming they have been deleted from Metashare since the last harvest). Such PIDs are then DELETEd from Metax based on the records' Metax UUID.
- Some test data was manipulated in order to utilize them better in tests.

kaisakui committed:
KP-7481 Strip spaces

kaisakui committed:
KP-7481 Add function to DELETE a dataset in Metax

kaisakui committed:
KP-7481 Add function to get PIDs in datacatalog

kaisakui committed:
KP-7481 Add function that deletes records in Metax

kaisakui committed:
KP-7481 Fix changed datacatalog id

kaisakui committed:
KP-7481 Change return statement to response

kaisakui committed:
KP-7481 Add tests for DELETE request

kaisakui committed:
KP-7481 Add tests for fetching pids

kaisakui committed:
KP-7481 Create class with constructor

kaisakui committed:
KP-7481 Add method to make requests with args

kaisakui committed:
KP-7481 GET request record id

kaisakui committed:
KP-7481 POST request new record

kaisakui committed:
KP-7481 PUT request of existing record

kaisakui committed:
KP-7481 DELETE request a record

kaisakui committed:
KP-7481 Get all pids in datacatalog

kaisakui committed:
KP-7481 Remove unused import

kaisakui committed:
KP-7481 Use MetaxAPI class

kaisakui committed:
KP-7481 Records_to_dict returns list of dicts

kaisakui committed:
KP-7581 Refactor how records are synced

kaisakui committed:
KP-7481 Move attributes to constructor

kaisakui committed:
KP-7481 Add comment

kaisakui committed:
KP-7481 Fix tests to check record id

kaisakui committed:
KP-7481 Rewrite test descriptions

kaisakui committed:
KP-7481 Add MetaxAPI class fixture

kaisakui committed:
KP-7481 Delete (now) redundant test

kaisakui committed:
KP-7481 Work with JSON not string

kaisakui committed:
KP-7481 Use JSON not string

kaisakui committed:
KP-7481 Fix test for POST

kaisakui committed:
KP-7481 Fix tests for PUT

kaisakui committed:
KP-7481 Return response, not None

kaisakui committed:
KP-7481 Fix DELETE tests and response checks

kaisakui committed:
KP-7481 Fix GET request of all pids

kaisakui committed:
KP-7481 Apply refactoring changes to tests

kaisakui committed:
KP-7481 Add tests to collect pids from APIs

kaisakui committed:
KP-7481 Apply fixes to more easily mock requests

kaisakui committed:
KP-7481 Fix mock

kaisakui committed:
KP-7481 Apply changes to main() in tests

kaisakui committed:
KP-7481 Apply mock change to tests

kaisakui committed:
KP-7481 Update readme

aajarven commented on README.md:
````diff
@@ -1,11 +1,26 @@
+
+
+### Requirements
+This tool works with Python 3.8 together with pip and virtualenv installed. The rest of the reuirements can be installed with ```pip install -r requirements.txt```.
````
reQuirements

aajarven replied:
fixed

aajarven commented on metadata_harvester_cli.py:
```diff
@@ -41,63 +41,112 @@ def last_harvest_date(filename):
-def send_data_to_metax(all_mapped_data_dict):
+    if not date_time:
+        metashare_records = api.fetch_records()
+    metashare_records = api.fetch_changed_records(date_time)
```
This looks suspicious. If `fetch_changed_records` works with `None` (and fetches all records), we don't need the `if` above. And if it doesn't work (crashes, fetches no records, or similar unwanted behaviour) then we are in trouble because we overwrite the good result from the if.

Gotta figure this out.

aajarven replied:
Unified the record fetching into one function that works with or without (`None`) timestamp.

aajarven commented on metax_api.py:
```diff
@@ -1,99 +1,122 @@
+                  "persistent_identifier": pid}
+        endpoint = "datasets"
+        result = self._make_request("GET", endpoint, params)
+        assert result is not None  # Searching with parameters will always return "count"
```
If this assertion ever produces an error, it's likely not going to be too intuitive to debug. Maybe a custom error instead (or if we are sure that this should never happen (e.g. we get a RequestException when making a request if there is no result)) we could just remove it.

aajarven replied:
Removed. We will get at least as good error message from the next line.

aajarven commented on tests/test_metadata_harvester_cli.py:
```diff
@@ -212,18 +193,19 @@ def test_send_data_to_metax_multiple_records(
-        "pid1": {"meta": "data", "id": "pid1"},
-        "pid2": {"infor": "mation", "id": "pid2"},
-    }
+    record_dict = [
```
This is no longer a dict, recommend renaming

aajarven replied:
Renamed this and other `record_dict`s too.

aajarven reviewed (commented):
Submitting a review as a todo-list for future me, please disregard.

aajarven committed:
KP-7481 typofix

aajarven committed:
KP-7481 Make the record fetching generator work consistently

Now the generator is always a generator (although sometimes iteration
ends immediately). This way we can iterate through the generator whether
records were found or not without having to explicitly check for None.

aajarven committed:
KP-7481 Unify record fetching functions

As the API is fine with `from=None`, we don't need two separate
functions for fetching updated/all records. Instead we can use one
function with default value None for the start date. This simplifies the
code a bit.

aajarven committed:
KP-7481 Remove assert from production code

The assert will not produce a specific catchable exception nor
informative error message, so it doesn't seem to add much value beyond
the TypeError we will get on the next line if the Metax ID is not found
(which should not happen in the first place). As it doesn't bring much
value, it was removed.

aajarven committed:
KP-7481 Rename variables to not imply that lists are dicts

The data type that `send_data_to_metax` expects is nowadays a list (that
contains lists), so the old variable names were somewhat misleading. Now
they no longer imply dictness.

aajarven committed:
KP-7481 Remove unnecessary guard clause

Even if there are no records, we should have at least an empty list,
meaning that iterating over it is safe anyways.

mmatthiesencsc commented:
Hi Anni, I reviewed your last 6 commits, thanks, looks good.
